### PR TITLE
feat(mcp): steer agents toward search_knowledge with precise routing rules

### LINKED ISSUE #324: feat(mcp): Steer agents toward search_knowledge — server instructions + tool description rewrites + soft errors
## Summary

MCP-connected agents (Claude Code in particular) systematically prefer `list_files` + `get_document` over `search_knowledge` for question-answering tasks. This issue implements the four-pattern reinforcement stack from the research brief [`/agent-experience/agent-rag-corpus-grounding-2026-05-22.md`](https://github.com/Destrayon/Connapse) in `connapse-product-engineering` (32KB, 2026-05-22), grounded in Anthropic engineering posts, the MCP spec 2025-06-18, and arXiv 2505.18135.

## Why

The research brief identifies three compounding root causes:

1. **Deterministic verbs beat probabilistic ones.** `list_files`/`get_document` read as "returns file X"; `search_knowledge` reads as "returns ranked chunks." arXiv 2505.18135 measured 20–40pp selection shifts from this kind of wording.
2. **Claude's filesystem-exploration training prior.** Claude Code is RL'd on Glob/Read/Grep; `list_files` + `get_document` is structurally identical, so the habit transfers. Must be overridden.
3. **No corpus-level prior.** Without a "what's in here" overview, enumeration *is* the landscape-survey behavior.

Concrete evidence in our code: [`McpTools.cs:175`](src/Connapse.Web/Mcp/McpTools.cs:175)'s `list_files` description currently reads *"Use to browse container contents before retrieving documents"* — actively promoting the dig pattern.

The brief argues description rewrites alone give only +5.85pp success and add 67% more steps. The four patterns must ship together to reinforce each other across four touchpoints per turn.

### P1 — Server-level `instructions` field
[`Program.cs:94-100`](src/Connapse.Web/Program.cs:94) currently sets only `ServerInfo`. Add `options.ServerInstructions = ...` with a directive like:

> For question-answering over container contents, always call `search_knowledge` first. Use `list_files` only when the user names a specific filename or asks for a file inventory. Use `get_document` only after `search_knowledge` returns a `DocumentId` you want to read in full.

SDK confirmed: `ModelContextProtocol.Server.McpServerOptions.ServerInstructions` exists in v1.0.0. Honored by Claude Code per dbreunig analysis; other clients vary.

### P2 — Prescriptive tool descriptions
Rewrite [`McpTools.cs`](src/Connapse.Web/Mcp/McpTools.cs) `[Description(...)]` attributes:
- `search_knowledge` → prefix with **"PREFERRED for question-answering."**
- `list_files` → **"Inventory only. DO NOT use to answer questions about content — call `search_knowledge` instead."**
- `get_document` → **"Use only after `search_knowledge` returns a specific `DocumentId` worth reading in full. For Q&A, prefer `search_knowledge`."**
- `container_list` → already returns descriptions; tighten phrasing to flag it as the entry point.

Aim for Linear MCP density (~750 description tokens per tool). Avoid long `Examples:` blocks (arXiv 2602.14878: no statistical improvement).

### P3 — Tool-output-as-hint (tip header on return values)
Prepend a tip line to `list_files` and `container_list` output. Example for `list_files`:

```
TIP: This container has 487 files (~3.2M tokens). For question-answering, call search_knowledge(query=\"...\") — it returns the relevant ~5 passages directly. Listing and reading individual files will exceed context.

Contents of /:
[DIR] docs/
...
```

Lands in the agent's most recent context window, so high efficacy regardless of whether ServerInstructions is honored.

### P11 — Token-budget soft-error returns
In `list_files`, when a container has many files (threshold TBD — start at, say, 50 docs or estimated 20k+ tokens of listing output), return an MCP error response with body:

> Listing N files would consume ~T tokens. Call `search_knowledge(query=\"...\")` instead, or pass `limit: 25` and `confirm_large: true` to proceed.

Requires adding `limit` and `confirm_large` parameters to `list_files`. Anthropic explicitly endorses error-as-teaching: errors are very salient to the agent.

## Files affected

- [`src/Connapse.Web/Program.cs`](src/Connapse.Web/Program.cs) — P1: set `ServerInstructions` (~5 lines)
- [`src/Connapse.Web/Mcp/McpTools.cs`](src/Connapse.Web/Mcp/McpTools.cs) — P2: rewrite descriptions across all 12 tools; P3: tip headers in `ListFiles` + `ContainerList`; P11: add `limit` / `confirm_large` params and soft-error path to `ListFiles`

## Out of scope (explicit non-goals)

- `container_describe` tool with LLM-generated corpus summary + entity inventory — separate ticket
- Renaming `list_files` → `list_files_inventory` (P4) — client-breaking, defer
- Auto-emitting `SKILL.md` per container at ingest — speculative, defer
- MCP Resources for discovery — see adjacent #155 (research brief argues against as a primary fix; Claude Code does not auto-attach Resources)
- Tool gating / `notifications/tools/list_changed` (P6)
- Parameter schema forcing function — `get_document` requiring an opaque handle from `search_knowledge` (P5)

## Test plan

- [ ] Unit: `search_knowledge` and `list_files` descriptions match expected strings
- [ ] Unit: `list_files` returns soft error (`isError: true`) when document count exceeds threshold and `confirm_large` is false; succeeds when `confirm_large: true` is passed
- [ ] Unit: `list_files` and `container_list` outputs contain expected tip-line prefix
- [ ] Manual: connect Claude Code, ask a question about a known container, verify it calls `search_knowledge` first (not `list_files` + `get_document`)
- [ ] Manual: verify `ServerInstructions` appears in the MCP `initialize` response over the wire (curl or MCP inspector)
- [ ] Existing MCP tool integration tests still pass

## Branch

`feature/<this-issue>-mcp-agent-steering`

## References

- Research brief: `connapse-product-engineering` container, `/agent-experience/agent-rag-corpus-grounding-2026-05-22.md`
- Anthropic, *Writing effective tools for AI agents* — https://www.anthropic.com/engineering/writing-tools-for-agents
- arXiv 2505.18135 — *Tool Preferences in Agentic LLMs are Unreliable*
- MCP spec 2025-06-18 — Lifecycle (instructions field)
- Adjacent: #155 (MCP Resources — different approach to same problem; brief argues against as primary fix)

Size estimate: **S–M** (50–150 lines: ~5 in Program.cs, ~30 description rewrites in McpTools.cs, ~50 for P3 + P11 changes in `ListFiles`).

### PR DESCRIPTION
**Pull request by Destrayon**
## Summary

Two-phase MCP agent-steering intervention to reduce `list_files` enumeration when `search_knowledge` is the right tool for content questions.

### v1 (commits 65c9c27 + 7e51675)
Prescriptive `ServerInstructions` + tool description rewrites: "call `search_knowledge` FIRST", "ENTRY POINT", "INVENTORY ONLY", "DO NOT enumerate files" imperatives. Benchmark verdict was net-positive on Claude Haiku 4.5 and Sonnet 4.6, but produced a documented failure mode on Opus 4.7: literal compliance to *competing* "FIRST" imperatives (one in `ServerInstructions`, one in the `container_list` description) collapsed `search_knowledge` adoption to zero on the `qa_factoid` bucket and inflated MCP call count by +41%.

### v2 (commits 9538f69 + a11825b + a834cb9 + 901dcc9 + 1d43e2c)
Restructured the intervention based on literature (Faghih et al. EMNLP 2025 on description-driven tool substitution; Anthropic's Opus 4.7 prompting guide on literal compliance) along three axes:

1. **`ServerInstructions` becomes the single source of routing voice** — four numbered scope-explicit conditional rules instead of unconditional imperatives. Re-licenses `search_knowledge` as the primary path when the container is already known (Rule 1), explicitly preventing the Opus literal-compliance chain.
2. **Tool `[Description]` attributes** rewritten as neutral functional text (one minimal disambiguation cue each, no routing imperatives).
3. **Inline TIPs in `container_list` and `list_files` output** are now conditional — emit only when an actual routing decision exists at the response. Trimmed from ~60-80 words to ~20-25.

The `list_files` soft-error path (when entries > 50 without `confirmLarge`) deliberately retains its routing imperative — that's a contextual policy refusal at the moment of violation, consistent with the v2 principle of "disambiguate output, don't re-teach routing."

### Out of scope
- Per-model description shipping
- `defer_loading` / Tool Search infrastructure (Connapse has 4 primary MCP tools, not dozens)
- Server-side per-MCP-tool-call telemetry
- Benchmark harness session-limit-detection fix
- Judge calibration
- Re-running the v1 benchmark (a v2 benchmark would validate the conditional-phrasing prediction)

### Design/plan docs
v2 spec and implementation plan live locally at `docs/superpowers/specs/2026-05-23-mcp-agent-steering-v2-design.md` and `docs/superpowers/plans/2026-05-23-mcp-agent-steering-v2.md` (not committed; happy to share).

## Test plan

- [x] All 683 unit tests pass (`dotnet test --filter "Category=Unit"`)
- [ ] Integration tests pass with Docker (`dotnet test --filter "Category=Integration"`)

## Follow-up validation (separate work)

The v2 benchmark should confirm:
- Opus `qa_factoid` `right_tool_first_rate` ≥ v1's +15.2pp
- Opus `qa_factoid` `mean_mcp_call_count` delta < +0.30 (v1 had +0.67)
- Opus `qa_factoid` `search_knowledge` adoption rate > 0 (v1 collapsed to 0)
- Haiku `qa_demo_realistic` `right_tool_first_rate` ≥ v1's +16.7pp (softer phrasing must not flatten)

Closes #324

## Summary by CodeRabbit

* **New Features**
  * Added optional parameters to file listing controls: `limit` to truncate results and `confirmLarge` to override size constraints.
  * Introduced conditional guidance tips for large folder inventories, redirecting to search when appropriate.

* **Improvements**
  * Refined tool descriptions to clarify agent workflows: container selection followed by search, then file inventory retrieval.
  * Enhanced file listing behavior with soft-limit enforcement and clearer truncation messaging.
  * Configured MCP server routing instructions for improved agent decision-making.

[Review Change Stack](https://app.coderabbit.ai/change-stack/Destrayon/Connapse/pull/325?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: src/Connapse.Web/Mcp/McpTools.cs [hunk 1]
```diff
@@ -222,27 +230,48 @@ public static async Task<string> ListFiles(
             }
         }
 
-        var text = $"Contents of {normalizedPath}:\n\n";
+        // Direct-child files (the only docs that actually render at this level)
+        var directChildDocs = documents
+            .Where(d => string.Equals(PathUtilities.GetParentPath(d.Path), normalizedPath, StringComparison.OrdinalIgnoreCase))
+            .ToList();
+
+        var visibleEntries = folderNames.Count + directChildDocs.Count;
```
[the remark this answers is not in the document]
Comment (Destrayon) [reply src/Connapse.Web/Mcp/McpTools.cs]: Fixed in 7e51675 — `int visibleEntries`.

File: src/Connapse.Web/Mcp/McpTools.cs [hunk 2]
```diff
@@ -222,27 +230,48 @@ public static async Task<string> ListFiles(
...
+            return $"Error: '{normalizedPath}' contains {visibleEntries} entries, which exceeds the soft listing limit of {ListFilesSoftLimit}. " +
+                   $"For question-answering, call `search_knowledge(query=\"...\", containerId=\"{containerId}\")` instead — it returns the relevant passages directly. " +
+                   $"If you genuinely need the full inventory, retry with `confirmLarge: true`, or paginate with `limit: 25`.";
+        }
+
+        var text = $"TIP: This listing contains {directChildDocs.Count} file(s) and {folderNames.Count} folder(s) at this level but NO file contents. To answer questions about what these files contain, call `search_knowledge(query=\"...\", containerId=\"{containerId}\")` instead — it returns the relevant passages directly without you having to read each file. Use this listing only if the user explicitly asked for an inventory or named a specific filename.\n\n";
+        text += $"Contents of {normalizedPath}:\n\n";
         var hasEntries = false;
+        var rendered = 0;
+        var effectiveLimit = limit ?? int.MaxValue;
```
[the remark this answers is not in the document]
Comment (Destrayon) [reply src/Connapse.Web/Mcp/McpTools.cs]: Fixed in 7e51675 — all three switched to explicit types (`bool hasEntries`, `int rendered`, `int effectiveLimit`).

File: tests/Connapse.Core.Tests/Mcp/McpToolsContainerListTests.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    public async Task ContainerList_EmptyReturnsPlainMessageWithoutTip()
+    {
+        _containerStore
+            .ListAsync(Arg.Any<int>(), Arg.Any<int>(), Arg.Any<CancellationToken>())
+            .Returns(new List<Container>());
+
+        var result = await McpTools.ContainerList(_services);
+
+        result.Should().Be("No containers found.");
+    }
```
[the remark this answers is not in the document]
Comment (Destrayon) [reply tests/Connapse.Core.Tests/Mcp/McpToolsContainerListTests.cs:34]: Pushing back on this one — the empty case is intentionally plain, not a regression of the steering goal.

The empty case means **zero containers exist in the system**. `search_knowledge` requires a `containerId` parameter, and there are no IDs to pass. Telling the agent to "try `search_knowledge`" when nothing is indexed sends it at a dead-end tool path — the right next action is for the user to create a container or upload data, not to search.

The PR's steering principle is: prefer `search_knowledge` over `list_files` / `get_document` **for content questions over real data**. That dichotomy doesn't exist in the zero-data boundary case — neither tool helps, so neither tool should be suggested. The test name `EmptyReturnsPlainMessageWithoutTip` documents this as deliberate design, and the non-empty test (`ContainerList_OutputBeginsWithTipPointingAtSearchKnowledge`) covers the real steering surface.

Keeping the empty-case assertion as-is.

**Comment by Destrayon** [comment]
## CodeRabbit v2 review — actions taken

**Fixed (Major):**
- `src/Connapse.Web/Mcp/McpTools.cs:196` — added upfront `limit > 0` guard in `ListFiles`. A non-positive `limit` would otherwise bypass the soft-error path (`!limit.HasValue` is false when limit is set) and the rendering loop would short-circuit on `rendered >= effectiveLimit`, producing misleading "(empty)" output for non-empty folders. Added `[Theory]` test covering `limit = 0, -1, -100`.

**Deferred (Nitpicks — test naming convention):**
- The 7 new tests in `McpServerConfigTests`, `McpToolsContainerListTests`, `McpToolsListFilesTests` use the 2-segment `Method_ConcatenatedScenarioResult` form to stay consistent with the existing 12+ v1 tests in `Mcp/` (e.g., `ListFiles_LargeListingReturnsSoftErrorByDefault`, `ListFiles_ConfirmLargeBypassesSoftError`).
- CodeRabbit's strict reading of CLAUDE.md (`MethodName_Scenario_ExpectedResult`, 3 segments) is correct, but applying it only to the 7 new tests would create inconsistency with the surrounding code. A project-wide test-naming cleanup is the right scope for this — out of v2's scope.

**Discussion-only thread (no code change):**
- `McpToolsContainerListTests.cs:34` — empty-case `Should().Be("No containers found.")` assertion. CodeRabbit and the author already converged in-thread (CodeRabbit acknowledged the v1 reasoning was correct: `search_knowledge` needs a `containerId`, so steering toward it in the empty case sends the agent to a dead-end). v2 preserves the agreed design. Leaving the GitHub thread open for context.

**Commit:** `6ba8e9b` on `feature/324-mcp-agent-steering`.

**Tests:** 390/390 Core unit tests pass (was 387 pre-fix; +3 for the new theory cases).
